### PR TITLE
[HA](Civo) Addition of HA cluster API system

### PR DESCRIPTION
**Pull request by dipankardas011**
Signed-off-by: Dipankar Das <dipankardas0115@gmail.com>

# Tasks / Issue :construction:
Issues
* #36 

# Solution :heavy_check_mark:
Create Instances then config them as req.

# Screenshots :framed_picture:

# Note to reviewers :notebook:

### CODE REVIEW

File: src/api/ha_civo/main.go [hunk 1]
```diff
@@ -77,25 +77,40 @@ func fetchAPIKey() string {
 // TODO demoScript try the script whether its working or not
 func demoScript() string {
 	return `#!/bin/bash
-sudo apt update -y
-sudo apt install nginx git -y
+sudo apt update -y 
+sudo apt install -y nginx git
 `
 }
```
Comment (dipankardas011) [inline src/api/ha_civo/main.go:83]: Using this function to get the PostInstall script to install stuff
Problem is its not working 😢 
does anyone know what to do

**Comment by dipankardas011** [comment]
Tasks
- [x] creation of instances
- [ ] load balance 
- [ ] config vms
- [ ] run single node k8s
- [ ] joined Worker Nodes
- [ ] joined additional control plane

**Comment by dipankardas011** [comment]
# ⚠️ [BUG] found in CIVO instance
scripts sent are not available for the instance to run so just have to wait for it to get resolved

**Comment by dipankardas011** [comment]
> #  [BUG] found in CIVO instance
> scripts sent are not available for the instance to run so just have to wait for it to get resolved

# (Temporary) Solution
what you want the script to do you can ssh into and execute the script as of now till the above issue gets resolved

File: src/api/ha_civo/main.go [hunk 2]
```diff
@@ -173,39 +172,40 @@ func CreateVM(name, region string) {
...
+		},
+		HostKeyCallback: hostKeyCallback,
+	}
+
+	// connect ot ssh server
+	conn, err := ssh.Dial("tcp", ip+":22", config)
+	if err != nil {
+		fmt.Println(err.Error())
+		return
+	}
```
Comment (dipankardas011) [inline src/api/ha_civo/main.go:205]: @AvineshTripathi there are some issue with this line
Thursday 24 November 2022 10:55:56 AM IST

**Comment by dipankardas011** [comment]
For now the workflow for development is
1. Put break point on `initkubeadmssh` func
2. Run debugger from ha-civo... main file in src/api
3. Once the debugger steps in the beard point
Execute / copy the preinitializatio script ssh into vm and execute it there after it's done successfully 
4. Continue with debugger 
5. If no err found in `initkubeadmssh` then comment down below

**Comment by dipankardas011** [comment]
### SSH config working 2/n task is complete
Friday 25 November 2022 11:46:55 PM IST

**Comment by dipankardas011** [comment]
# Proposal of change
`k8s` -> `k3s`

**Comment by dipankardas011** [comment]
as we don't need initialiing script any more bug can be removed

**Comment by dipankardas011** [comment]
### Some of the useful links
- [Docs of k3s](https://docs.k3s.io/installation/ha)
- [Blog](https://gabrieltanner.org/blog/ha-kubernetes-cluster-using-k3s/)
- [HA proxy](http://docs.haproxy.org/2.7/intro.html)

**Comment by dipankardas011** [comment]
from here development(coding) starts
---

**Comment by dipankardas011** [comment]
### [BUG] found 🐞 
https://github.com/kubesimplify/ksctl/blob/bdf085cac3c75beff523a038502be101a4cd4fe1/src/api/ha_civo/loadbalancer.go#L36
semicolon was missing in

**Comment by dipankardas011** [comment]
VMs are done ✔️ 
Save the instanceIDs 
delete VMs using the saved state

File: src/api/ha_civo/controlplane.go [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	diskImg, err := client.GetDiskImageByName("ubuntu-focal")
+	if err != nil {
+		return nil, err
+	}
+
+	name := fmt.Sprintf("%s-ksctl-cp-%d", clusterName, number)
```
Comment (AvineshTripathi) [inline src/api/ha_civo/controlplane.go:62]: confirm if going with randomized naming convention a better way inplace of sequence

if we are going with random make sure to remove parameters `int`

**Review by AvineshTripathi — COMMENTED** [review]
@kranurag7 please review the scripts running

**Comment by dipankardas011** [comment]
## NOTE
- [ ] Final check using the `src/api/ha-civo_dev.go ` which will be removed 
- [ ] Check of all the rest of the civo and local providers are working
- [ ] `Squash and merge` when merging this PR 

cc @AvineshTripathi @kranurag7 @dipankardas011

**Comment by dipankardas011** [comment]
> ## NOTE
> * [ ]  Final check using the `src/api/ha-civo_dev.go ` which will be removed
> * [ ]  Check of all the rest of the civo and local providers are working
> * [ ]  `Squash and merge` when merging this PR
> 
> cc maintainers @AvineshTripathi @kranurag7 @dipankardas011

final check for any fix of `Use of insecure HostKeyCallback implementation` as the codeql ❌

**Pull request merged**